crates/wasi-common/wig/src/raw_types.rs

```diff
@@ -179,86 +179,76 @@ fn builtin_tokens(mode: Mode, builtin: witx::BuiltinType) -> TokenStream {
+        witx::TypeRef::Value(v) => match &**v {
+            witx::Type::Builtin(b) => builtin_tokens(mode, *b),
+            witx::Type::Pointer(pointee) => {
+                let pointee = tref_tokens(mode, pointee);
+                match mode {
+                    Mode::Host => quote!(*mut #pointee),
+                    Mode::Wasi => panic!("pointers have target-specific size"),
+                    Mode::Wasi32 => quote!(u32),
+                }
             }
-        }
-        witx::DatatypeIdent::ConstPointer(pointee) => {
-            let pointee = ident_tokens(mode, pointee);
-            match mode {
-                Mode::Host => quote!(*const #pointee),
-                Mode::Wasi => panic!("pointers have target-specific size"),
-                Mode::Wasi32 => quote!(u32),
+            witx::Type::ConstPointer(pointee) => {
+                let pointee = tref_tokens(mode, pointee);
+                match mode {
+                    Mode::Host => quote!(*const #pointee),
+                    Mode::Wasi => panic!("pointers have target-specific size"),
+                    Mode::Wasi32 => quote!(u32),
+                }
             }
-        }
-        witx::DatatypeIdent::Array(element) => {
-            let element_name = ident_tokens(mode, element);
-            match mode {
-                Mode::Host => quote!((*const #element_name, usize)),
-                Mode::Wasi => panic!("arrays have target-specific size"),
-                Mode::Wasi32 => quote!((u32, u32)),
+            witx::Type::Array(element) => {
+                let element_name = tref_tokens(mode, element);
+                match mode {
+                    Mode::Host => quote!((*const #element_name, usize)),
+                    Mode::Wasi => panic!("arrays have target-specific size"),
+                    Mode::Wasi32 => quote!((u32, u32)),
+                }
             }
-        }
+            t => panic!("cannot give name to anonymous type {:?}", t),
+        },
     }
 }
 
 /// Test whether the given struct contains any union members.
-fn struct_has_union(doc: &witx::Document, s: &witx::StructDatatype) -> bool {
-    s.members.iter().any(|member| match &member.type_ {
-        witx::DatatypeIdent::Ident(ident) => match &doc.datatype(&ident.name).unwrap().variant {
-            witx::DatatypeVariant::Union(_) => true,
-            witx::DatatypeVariant::Struct(s) => struct_has_union(doc, &s),
-            _ => false,
-        },
+fn struct_has_union(s: &witx::StructDatatype) -> bool {
+    s.members.iter().any(|member| match &*member.tref.type_() {
+        witx::Type::Union { .. } => true,
+        witx::Type::Struct(s) => struct_has_union(&s),
         _ => false,
     })
 }
 
+fn tref_has_target_size(tref: &witx::TypeRef) -> bool {
+    match tref {
+        witx::TypeRef::Name(nt) => namedtype_has_target_size(&nt),
+        witx::TypeRef::Value(t) => type_has_target_size(&t),
+    }
+}
+
 /// Test whether the given type has a target-specific size.
```

Comment:
This comment could say "named type" instead of "type", to distinguish it from the comment below.